Adapt groupapply to new DataFrames

piever opened the pull request:
Here are the changes for `groupapply` to work with the new DataFrames design. I simply check what columns will be used to produce the plot, exclude the datapoints where there's missing data in one of those columns, convert to a dataframe of arrays and then proceed as usual with the analysis and plotting.

mkborregaard commented on src/groupederror.jl:
```diff
@@ -202,14 +204,24 @@ Output is a `GroupedError` with error computed according to the keyword `compute
+                    :discrete : :continuous
     end
+
+    # Convert to dataframe with only relevant columns and no missing data
```
So far, we have not yet to decided what to do when the DataFrame has missing values. I think it is better to wait with this and error on Nulls, to have consistency across the repo. Could you take this out, I will merge the rest and we can put this bit in an un-merged PR.

mkborregaard commented:
Great work! I have a single comment on not at this point to automatically remove Nulls.

nalimilan commented on src/groupederror.jl:
```diff
@@ -190,10 +189,13 @@ function get_groupederror(trend,variation, f, df::AbstractDataFrame, axis_type,
 
     groupapply(f::Function, df, args...;
-                axis_type = :auto, compute_error = :none, group = [],
+                axis_type = :auto, compute_error = :none, group = Array{Symbol,1}(0),
```
`Symbol[]` is shorter (same below).

nalimilan commented on src/groupederror.jl:
```diff
@@ -202,14 +204,24 @@ Output is a `GroupedError` with error computed according to the keyword `compute
+
     if !(axis_type in [:discrete, :continuous])
-        axis_type = (typeof(df[args[1]])<:PooledDataArray) ? :discrete : :continuous
+        axis_type = (typeof(dfn[args[1]])<:Union{CategoricalArray,NullableCategoricalArray}) ?
```
Shouldn't `Array{String}` also be considered as categorical?

mkborregaard replied:
Yes, I just had the same thought in response to the comment here: https://github.com/JuliaPlots/StatPlots.jl/pull/34#issuecomment-274056998
In StatPlots, both `Array{String}` and `Array{Symbol}` are automatically treated as categorical variables.

piever replied:
I totally agree, that's done [here](https://github.com/JuliaPlots/StatPlots.jl/blob/master/src/groupederror.jl#L213): if there's an Array (not Categorical) of not numbers (e.g. strings or symbols) and the user didn't set `axis_type = :discrete` I give a warning and convert do discrete axis (if this is standard policy in StatPlots maybe I can remove the warning). Now the two checks (is array categorical and is `eltype<:Real`) got separated because I thought it's easier to check for the eltype after converting to normal arrays, but it's easier to check whether a column is categorical before converting.

nalimilan replied:
I think it makes more sense to treat as categorical without a warning when the user didn't expliciltly request a continuous axis.

mkborregaard replied:
yes

nalimilan commented on src/groupederror.jl:
```diff
@@ -202,14 +204,24 @@ Output is a `GroupedError` with error computed according to the keyword `compute
+                vectorify(compute_error[2]) : Array{Symbol,1}(0)
+    rel_cols = filter(t -> (t in names(dfn)), union([args...], err_col, vectorify(group)))
+    df = dfn[complete_cases(dfn[:, rel_cols]),rel_cols]
+    df = DataFrame(convert.([Array],getindex.([df],rel_cols)) , rel_cols)
```
Wouldn't `Array.(...)` work here? Can't you also do `df[rel_cols]` directly?

piever replied:
`df[rel_cols]` won't work, because you need an array of columns for the dot syntax to work, whereas `df[rel_cols]` returns a dataframe. Then the array of column becomes an array of arrays, which is the only way to get a dataframe without nullable arrays AFAIK. `Array.(...)` is much more legible than what I have and it does just the same, I'll change to that.

nalimilan replied:
Probably more readable: `DataFrame([Array(df[col]) for col in rel_cols], rel_cols)` or `DataFrame(; (col=>Array(df[col]) for col in rel_col)...)`.

nalimilan commented on src/groupederror.jl:
```diff
@@ -1,15 +1,14 @@
 
 function extend_axis(xsmall, ysmall, xaxis, val)
-    df = DataFrame(x = xsmall, y = ysmall)
+    df = DataFrame(Any[xsmall,ysmall], [:x,:y])
```
Do you really need to use this special constructor here?

piever replied:
I think I do: if I don't use it, it may happen that I try and join a dataframe of nullablearrays with a dataframe of normal arrays, which gives an error. The approach of `groupapply` is to convert what the user gave to a df of only arrays and then only work with that type of dataframes.

nalimilan replied:
Can you file a bug about this in DataFrames?

piever replied:
Sure!

piever replied:
Thanks for taking care of the bug. Still, I'll probably keep this somewhat uglier version that works no matter what until NullableArrays stabilizes, and then I'll go over the code again to see if it can be made more elegant.

mkborregaard replied:
That is OK

piever commented:
I should've addressed all of the concerns. Changing from this to excluding rows with missing data is actually one line so we can change it back anytime very easily.

mkborregaard commented:
Alright, I will merge it tomorrow unless @nalimilan or @tbreloff have further comments here. Thanks again for the great work!

tbreloff commented:
No comments (though I admit to not looking at it closely)

On Wed, Jan 25, 2017 at 1:39 PM Michael Krabbe Borregaard <
notifications@github.com> wrote:

> Alright, I will merge it tomorrow unless @nalimilan
> <https://github.com/nalimilan> or @tbreloff <https://github.com/tbreloff>
> have further comments here. Thanks again for the great work!
>
>
>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaPlots/StatPlots.jl/pull/36#issuecomment-275240930>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA492tfaGtIgdF59323YcRobcYwS6kCaks5rV8DugaJpZM4LqFAT>
> .
>

mkborregaard commented:
Great!
